translator/trace/zipkin/zipkinv1_to_protospan.go

```diff
@@ -481,3 +485,14 @@ func (ep *endpoint) createAttributeMap() map[string]string {
 	}
 	return attributeMap
 }
+
+func SetTimestampsIfUnset(span *tracepb.Span) {
+	// zipkin allows timestamp to be unset, but opentelemetry-collector expects it to have a value.
+	// If this is unset, the conversion from open census to the internal trace format breaks
+	// what should be an identity transformation oc -> internal -> oc
+	if span.StartTime == nil {
+		now := internal.TimeToTimestamp(time.Now())
+		span.StartTime = now
+		span.EndTime = now
+	}
```

Comment:
I'm not sure about this but I don't know what the right thing to do is. Addin/modifying data like this just to make serialize happy doesn't sound that great to me. This mean backends will show misleading information to users.

Alternatively, we could just drop such spans as a validation step in the zipkin receiver and log messages but we can't do that as zipkin spec allows for the  possibility of not having timestamps. Is there a way we can serialize it differently without adding fake timestamps?

If there is no other viable solution, we should at least add an attribute to the span that represents that the timestamps are fake and shouldn't be trusted. Also, can we may be use more obvious fake value for timestamp like unix epoch or something?

Reply:
Zipkin spec on this:

```
Epoch microseconds of the start of this span, possibly absent if
incomplete.

For example, 1502787600000000 corresponds to 2017-08-15 09:00 UTC

This value should be set directly by instrumentation, using the most
precise value possible. For example, gettimeofday or multiplying epoch
millis by 1000.

There are three known edge-cases where this could be reported absent.

    A span was allocated but never started (ex not yet received a timestamp)
    The span’s start event was lost
    Data about a completed span (ex tags) were sent after the fact

```

Reply:
-     A span was allocated but never started (ex not yet received a timestamp)

If this was the only edge-case, we could just drop the span as it didn't actually represent a real operation but we have two more scenarios that can result in spans without timestamps so we can't just drop.

It looks like setting a static timestamp that is obviously fake along with a well-known attribute that says the timestamps were injected by Otel collector as a workaround is probably our best bet. Does Jaeger proto allow having empty timestamps? If so, an OTLP>Jaeger/Zipkin translator should probably set the timestamp to nil again when it encounters this tag on a span. One concern with setting a very old timestamp is that it might affect some backends in an unexpected way. For example, some backend might drop the span very soon if it stores only recent data and by span operation timestamps instead of ingestion timestamp.

Reply:
I'm not familiar with Jaeger, will need someone else to weigh in on that

Reply:
> It looks like setting a static timestamp that is obviously fake along with a well-known attribute that says the timestamps were injected by Otel collector as a workaround is probably our best bet. Does Jaeger proto allow having empty timestamps? If so, an OTLP>Jaeger/Zipkin translator should probably set the timestamp to nil again when it encounters this tag on a span. One concern with setting a very old timestamp is that it might affect some backends in an unexpected way. For example, some backend might drop the span very soon if it stores only recent data and by span operation timestamps instead of ingestion timestamp.

I agree with this approach. We can set a fake current timestamp along with an attribute identifying this workaround and use it in Jaeger/zipkin exporters to reconstruct null timestamps.

Reply:
Suggestions on the attribute name? These exist already https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/zipkin/attributekeys.go#L19

Reply:
Probably `zipkin.startTime.absent`. @owais what do you think?

Reply:
That sounds like a zipkin convention. We should probably have an otel convention about such tags. May be `otel.zipkin.missingTimestamp` or `otel.zipkin.missingFields` with value being comma separated list of field names?  May be we should have otel collector specific conventions for behavior like this. Perhaps prefix with `otelcol` or something?

> We can set a fake current timestamp
Why current timestamp? Can we do zero time? Like 0 seconds since unix epoch? I'm afraid if we use current time and the span get exported using something other than zipkin, it might be very misleading to users when seeing the span in a tracing backend. "zero time" will at least raise some eye brows.

Reply:
@owais , 0 timestamps look cleaner, but if I understood your concern right it doesn't look like the best way:

> One concern with setting a very old timestamp is that it might affect some backends in an unexpected way. For example, some backend might drop the span very soon if it stores only recent data and by span operation timestamps instead of ingestion timestamp.

IMO still dropping spans with 0 timestamp looks like the best way to handle it in otel-collector. @chris-smith-zocdoc could you please let us know which backend you use and if the spans with no timespans are actually visible or they needed for aggregation purposes?

Reply:
> IMO still dropping spans with 0 timestamp looks like the best way to handle it in otel-collector

I disagree. Throwing away data seems suboptimal. 

Through coincidence they actually end up as a zero timestamp right now, but after the translation from OC -> internal -> OC for the zipkin export, the timestamp is becoming invalid and causing the serialization issue.  

> could you please let us know if the spans with no timestamps are actually visible in the backend that you use or they needed for aggregation purposes?

Yes they're visible. We have two backends, one is zipkin the other is honeycomb.

The data in these spans are just events and annotations, sent after the original span. This appears to be a common behavior in the zipkin libraries we're using for javascript and scala.
```
[
    {
        "traceId": "404059d49fac24fc",
        "id": "404059d49fac24ff",
        "annotations": [
            {
                "timestamp": 1589292979495000,
                "value": "zipkin-js.flush"
            }
        ],
        "tags": {
            "response_size_bytes": "9622"
        }
    }
]
```

In zipkin the data is merged into the original span. 

In honeycomb these end up as a second event in their system. I've considered building a buffer/merge processor (similar to tail sampling) to do the merge in OTel before sending it to them.

Reply:
Agree that throwing away the spans does not sound correct for people who might be using zipkin end to end (or even some parts of it) since it is a valid zipkin use case. I still have concerns about zero'ed timestamps causing backends to drop spans too soon. May be backends can use the tag we inject to detect this edge-case but not sure we can expect all backends to do this especially since it is not part of the otel spec.  I guess adding current timestamp for both start and end is the safest choice here even though it's not very optimal for backends other than zipkin.

Reply:
Sounds good. Let's stick with the current timestamp. 

Also @owais had a good point about about the attribute name. We probably should not use a name that looks like zipkin convention. @owais, what about `otel.zipkin.absentField.startTime`?